File: packages/core/src/extensions/builtin/statusline-themes/themes.ts
/**
 * Theme definitions for the statusline-themes extension. A theme recolors the
 * *already-rendered* status-line rows (whatever the active layout produced): we
 * strip the existing ANSI first, then repaint the plain text our way. ANSI
 * helpers live in ./ansi so the layout code can share them.
 */
import { fg, hue, type RGB, sampleStops, stripAnsi } from "./ansi";

export type ThemeId =
    | "default"
    | "mono"
    | "matrix"
    | "ocean"
    | "sunset"
    | "synthwave"
    | "fire"
    | "rainbow"
    | "heat"
    | "neon"
    | "gold"
    | "cyber";

type ThemeKind =
    | { kind: "off" } // leave the native colored render untouched
    | { kind: "solid"; color: RGB } // one flat color
    | { kind: "gradient"; stops: RGB[] } // left→right interpolation across 2+ stops
    | { kind: "rainbow" }; // per-character hue sweep

export interface Theme {
    id: ThemeId;
    label: string;
    description: string;
    spec: ThemeKind;
}

export const THEMES: Theme[] = [
    {
        id: "default",
        label: "default",
        description: "the native colors (agent orange, model cyan, cost green)",
        spec: { kind: "off" },
    },
    {
        id: "mono",
        label: "mono",
        description: "no color — plain monochrome text",
        spec: { kind: "solid", color: { r: 170, g: 170, b: 170 } },
    },
    {
        id: "matrix",
        label: "matrix",
        description: "all green, terminal-hacker vibes",
        spec: { kind: "solid", color: { r: 0, g: 255, b: 102 } },
    },
    {
        id: "ocean",
        label: "ocean",
        description: "blue → cyan gradient",
        spec: {
            kind: "gradient",
            stops: [
                { r: 36, g: 92, b: 255 },
                { r: 0, g: 224, b: 224 },
            ],
        },
    },
    {
        id: "sunset",
        label: "sunset",
        description: "orange → magenta gradient",
        spec: {
            kind: "gradient",
            stops: [
                { r: 255, g: 153, b: 51 },
                { r: 224, g: 32, b: 160 },
            ],
        },
    },
    {
        id: "synthwave",
        label: "synthwave",
        description: "magenta → cyan gradient (retro neon)",
        spec: {
            kind: "gradient",
            stops: [
                { r: 255, g: 41, b: 184 },
                { r: 41, g: 224, b: 255 },
            ],
        },
    },
    {
        id: "fire",
        label: "fire",
        description: "red → yellow gradient",
        spec: {
            kind: "gradient",
            stops: [
                { r: 255, g: 40, b: 0 },
                { r: 255, g: 214, b: 0 },
            ],
        },
    },
    {
        id: "rainbow",
        label: "rainbow",
        description: "every character a different hue 🌈",
        spec: { kind: "rainbow" },
    },
    {
        id: "heat",
        label: "heat",
        description: "green → yellow → red, the context-bar heatmap",
        spec: {
            kind: "gradient",
            // AKCodez context-bar palette (gist ffb420ba…): 0-50% green→yellow, 50-100% yellow→red.
            stops: [
                { r: 0, g: 200, b: 80 },
                { r: 220, g: 200, b: 0 },
                { r: 220, g: 40, b: 20 },
            ],
        },
    },
    {
        id: "neon",
        label: "neon",
        description: "yellow → magenta → cyan (AKCodez daily-driver)",
        spec: {
            kind: "gradient",
            // The daily-driver element colors: yellow repo, magenta model, cyan branch.
            stops: [
                { r: 255, g: 255, b: 0 },
                { r: 255, g: 0, b: 255 },
                { r: 0, g: 184, b: 220 },
            ],
        },
    },
    {
        id: "gold",
        label: "gold",
        description: "solid gold (#FFFF00)",
        spec: { kind: "solid", color: { r: 255, g: 255, b: 0 } },
    },
    {
        id: "cyber",
        label: "cyber",
        description: "solid cyan (#00B8DC)",
        spec: { kind: "solid", color: { r: 0, g: 184, b: 220 } },
    },
];

export const DEFAULT_THEME: ThemeId = "default";

export function getTheme(id: string | undefined): Theme {
    return THEMES.find((t) => t.id === id) ?? THEMES[0];
}

/**
 * Recolor one rendered row according to a theme. `off` returns the row
 * unchanged (keep native colors). Everything else strips the existing ANSI and
 * repaints the plain text. Width is preserved (ANSI adds no visible columns).
 */
export function applyTheme(line: string, theme: Theme): string {
    if (theme.spec.kind === "off") return line;
    const plain = stripAnsi(line);
    if (plain.length === 0) return line;

    if (theme.spec.kind === "solid") {
        return fg(theme.spec.color, plain);
    }

    // Per-character coloring for gradient / rainbow.
    const chars = [...plain];
    const n = Math.max(1, chars.length - 1);
    let out = "";
    for (let i = 0; i < chars.length; i++) {
        const t = i / n;
        const color = theme.spec.kind === "rainbow" ? hue(t * 320) : sampleStops(theme.spec.stops, t);
        out += fg(color, chars[i]);
    }
    return out;
}

File: internal/modules/core/extension/statusline/themes.go
package statusline

// Colour themes recolour the ALREADY-RENDERED rows, whatever the active
// layout produced: the existing colour is stripped and the plain text is
// repainted. That is what lets the two axes compose — a layout decides the
// structure, a theme decides the colour, and neither knows about the other.

// themeKind is how a theme paints.
type themeKind int

const (
	// themeOff leaves the native colours alone.
	themeOff themeKind = iota
	// themeSolid is one flat colour.
	themeSolid
	// themeGradient interpolates left to right across two or more stops.
	themeGradient
	// themeRainbow sweeps the hue per character.
	themeRainbow
)

// Theme is one colour preset.
type Theme struct {
	ID          string
	Description string
	kind        themeKind
	colour      RGB
	stops       []RGB
}

// Themes is every colour preset, in menu order.
var Themes = []Theme{
	{ID: "default", Description: "the native colours (agent orange, model cyan, cost green)", kind: themeOff},
	{ID: "mono", Description: "no colour — plain monochrome text", kind: themeSolid, colour: RGB{170, 170, 170}},
	{ID: "matrix", Description: "all green, terminal-hacker vibes", kind: themeSolid, colour: RGB{0, 255, 102}},
	{ID: "ocean", Description: "blue → cyan gradient", kind: themeGradient,
		stops: []RGB{{36, 92, 255}, {0, 224, 224}}},
	{ID: "sunset", Description: "orange → magenta gradient", kind: themeGradient,
		stops: []RGB{{255, 153, 51}, {224, 32, 160}}},
	{ID: "synthwave", Description: "magenta → cyan gradient (retro neon)", kind: themeGradient,
		stops: []RGB{{255, 41, 184}, {41, 224, 255}}},
	{ID: "fire", Description: "red → yellow gradient", kind: themeGradient,
		stops: []RGB{{255, 40, 0}, {255, 214, 0}}},
	{ID: "rainbow", Description: "every character a different hue", kind: themeRainbow},
	{ID: "heat", Description: "green → yellow → red, the context-bar heatmap", kind: themeGradient,
		stops: []RGB{{0, 200, 80}, {220, 200, 0}, {220, 40, 20}}},
	{ID: "neon", Description: "yellow → magenta → cyan", kind: themeGradient,
		stops: []RGB{{255, 255, 0}, {255, 0, 255}, {0, 184, 220}}},
	{ID: "gold", Description: "solid gold", kind: themeSolid, colour: RGB{255, 255, 0}},
	{ID: "cyber", Description: "solid cyan", kind: themeSolid, colour: RGB{0, 184, 220}},
}

// DefaultTheme leaves the native colours alone.
const DefaultTheme = "default"

// GetTheme finds a theme by id, falling back to the default.
func GetTheme(id string) Theme {
	for _, t := range Themes {
		if t.ID == id {
			return t
		}
	}
	return Themes[0]
}

// Apply recolours one rendered row.
//
// Width is preserved because colour escapes occupy no columns — which is the
// property that lets this run AFTER the layout has already fitted its content
// to the terminal.
func (t Theme) Apply(line string) string {
	if t.kind == themeOff {
		return line
	}
	plain := StripANSI(line)
	if plain == "" {
		return line
	}
	if t.kind == themeSolid {
		return FG(t.colour, plain)
	}

	chars := []rune(plain)
	n := float64(max(1, len(chars)-1))
	var out string
	for i, r := range chars {
		at := float64(i) / n
		colour := SampleStops(t.stops, at)
		if t.kind == themeRainbow {
			colour = Hue(at * 320)
		}
		out += FG(colour, string(r))
	}
	return out
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
